t/re-gmatch.t

```diff
@@ -0,0 +1,451 @@
+                    ngx.say(m[3])
+                end
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+hello!
+hello
+false
+!
+world!
+world
+false
+!
+--- error_log eval
+qr/\[TRACE\s+\d+\s+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 9: unmatched trailing captures are false
+--- http_config eval: $::HttpConfig
+--- config
+    location = /re {
+        content_by_lua_block {
+            local iter = ngx.re.gmatch("hello", [[(\w+)(, .+)?(!)?]], "jo")
+            if iter then
+                while true do
+                    local m = iter()
+                    if not m then
+                        return
+                    end
+                    ngx.say(m[0])
+                    ngx.say(m[1])
+                    ngx.say(m[2])
+                    ngx.say(m[3])
+                end
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+hello
+hello
+false
+false
+--- error_log eval
+qr/\[TRACE\s+\d+\s+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: unmatched name captures are false
+--- http_config eval: $::HttpConfig
+--- config
+    location = /re {
+        content_by_lua_block {
+            local iter = ngx.re.gmatch(
+                "hello! world!",
+                [[(?<first>\w+)(?<second>, .+)?(?<third>!)]], "jo")
+            if iter then
+                while true do
+                    local m = iter()
+                    if not m then
+                        return
+                    end
+                    ngx.say(m[0])
+                    ngx.say(m[1])
+                    ngx.say(m[2])
+                    ngx.say(m[3])
+                    ngx.say(m.first)
+                    ngx.say(m.second)
+                    ngx.say(m.third)
+                end
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+hello!
+hello
+false
+!
+hello
+false
+!
+world!
+world
+false
+!
+world
+false
+!
+--- error_log eval
+qr/\[TRACE\s+\d+\s+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 11: subject is not a string type
+--- http_config eval: $::HttpConfig
+--- config
+    location /re {
+        content_by_lua_block {
+            local iter = ngx.re.gmatch(120345, "[1-9]+", "jo")
+            local m1 = iter()
+            local m2 = iter()
+            ngx.say(m1[0])
+            ngx.say(m2[0])
+        }
+    }
+--- request
+GET /re
+--- response_body
+12
+345
+--- no_error_log
+[error]
+attempt to get length of local 'subj' (a number value)
+
+
+
+=== TEST 12: an exhausted gmatch iterator should return nil
+--- http_config eval: $::HttpConfig
+--- config
+    location = /re {
+        content_by_lua_block {
+            local iter = ngx.re.gmatch("hello", [[\w+]])
+            local m = iter()
+            ngx.say("matched: ", m[0])
+            ngx.say("matched: ", iter())
+            ngx.say("matched: ", iter())
+        }
+    }
+--- request
+GET /re
+--- response_body
+matched: hello
+matched: nil
+matched: nil
+--- no_error_log
+[error]
```

Comment:
I was under the impression that this library's tests were supposed to test the JIT capabilities of the Lua code only, although, I do see that most of those tests come from `re-match.t`. No opinion on the matter, just saying.

However, what those tests are missing I believe, are test cases related to potential errors, as can be found in `035-gmatch.t` of the lua-nginx-module test suite. Errors from `re_match_compile` and `lua_ffi_exec_regex`. Should we add those, or will they be run from the lua-nginx-module test suite with this implementation? Quid if one does not run those tests from lua-nginx-module?